tests/common/hashkv_test.go

```diff
@@ -0,0 +1,248 @@
+						}
+						ctx, cancel := context.WithTimeout(t.Context(), 30*time.Second)
+						defer cancel()
+
+						clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+						defer clus.Close()
+
+						cc := testutils.MustClient(clus.Client())
+						tombstoneRevs, latestRev := populateDataForHashKV(t, cc, keys)
+
+						compactedOnRev := tombstoneRevs[0]
+
+						// If compaction revision isn't a tombstone, select a revision in the middle of two tombstones.
+						if !compactedOnTombstoneRev {
+							require.Greater(t, len(tombstoneRevs), 1)
+							compactedOnRev = (tombstoneRevs[0] + tombstoneRevs[1]) / 2
+							require.Greater(t, compactedOnRev, tombstoneRevs[0])
+							require.Greater(t, tombstoneRevs[1], compactedOnRev)
+						}
+
+						_, err := cc.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+						require.NoError(t, err)
+
+						for rev := compactedOnRev; rev <= latestRev; rev++ {
+							verifyConsistentHashKVAcrossAllMembers(t, cc, rev)
+						}
+					})
+				}
+			})
+		}
+	}
+}
+
+// TestVerifyHashKVAfterTwoCompactionsOnTombstone tests that HashKV is consistent
+// across all members after two physical compactions on tombstone revisions.
+func TestVerifyHashKVAfterTwoCompactionsOnTombstone(t *testing.T) {
+	testRunner.BeforeTest(t)
+	for _, tc := range clusterTestCases() {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.config.ClusterSize < 2 {
+				t.Skip("Skipping test for single-member cluster")
+			}
+			ctx, cancel := context.WithTimeout(t.Context(), 30*time.Second)
+			defer cancel()
+
+			clus := testRunner.NewCluster(ctx, t, config.WithClusterConfig(tc.config))
+			defer clus.Close()
+
+			cc := testutils.MustClient(clus.Client())
+			tombstoneRevs, latestRev := populateDataForHashKV(t, cc, []string{"key0"})
+			require.GreaterOrEqual(t, len(tombstoneRevs), 2)
+
+			firstCompactOnRev := tombstoneRevs[0]
+			t.Logf("COMPACT rev=%d", firstCompactOnRev)
+			_, err := cc.Compact(ctx, firstCompactOnRev, config.CompactOption{Physical: true})
+			require.NoError(t, err)
+
+			secondCompactOnRev := tombstoneRevs[1]
+			t.Logf("COMPACT rev=%d", secondCompactOnRev)
+			_, err = cc.Compact(ctx, secondCompactOnRev, config.CompactOption{Physical: true})
+			require.NoError(t, err)
+
+			for rev := secondCompactOnRev; rev <= latestRev; rev++ {
+				verifyConsistentHashKVAcrossAllMembers(t, cc, rev)
+			}
+		})
+	}
+}
+
+// TestVerifyHashKVAfterCompactionOnLastTombstone tests that HashKV is consistent
+// across all members after a physical compaction on the last tombstone revision.
+func TestVerifyHashKVAfterCompactionOnLastTombstone(t *testing.T) {
+	testRunner.BeforeTest(t)
+
+	for _, keys := range [][]string{
+		{"key0"},
+		{"key0", "key1"},
+	} {
+		t.Run(fmt.Sprintf("Keys=%v", keys), func(t *testing.T) {
+			for _, tc := range clusterTestCases() {
+				t.Run(tc.name, func(t *testing.T) {
+					if tc.config.ClusterSize < 2 {
+						t.Skip("Skipping test for single-member cluster")
+					}
```

Comment:
The test scope is across cluster members, so it can be skipped in the single-member case.